pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1242,43 +1233,102 @@ func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string
+				" The policy parameters will work only with VSAN Datastore."+
+				" So, please specify a valid VSAN datastore", datastore)
+		}
 
-	// Create a virtual disk manager
-	virtualDiskManager := object.NewVirtualDiskManager(vs.client.Client)
+		// 1. Create dummy VM with the disk attached.
+		dummyVMName, err := vs.createDummyVMWithDiskAttached(ctx, dc, ds, volumeOptions)
+		if err != nil {
+			return "", err
+		}
 
-	// Create specification for new virtual disk
-	vmDiskSpec := &types.FileBackedVirtualDiskSpec{
-		VirtualDiskSpec: types.VirtualDiskSpec{
-			AdapterType: LSILogicControllerType,
-			DiskType:    diskFormat,
-		},
-		CapacityKb: int64(volumeOptions.CapacityKB),
-	}
+		vmRegex := vs.cfg.Global.WorkingDir + dummyVMName
+		dummyVM, err := f.VirtualMachine(ctx, vmRegex)
+		if err != nil {
+			return "", err
+		}
+		vmDiskPath, err := getVMDiskPath(vs.client, dummyVM, volumeOptions.Name+".vmdk")
+		if err != nil {
+			deleteVM(ctx, dummyVM)
+			return "", fmt.Errorf("Failed to get disk path for disk on VM: %q with err: %+v", dummyVMName, err)
+		}
+		if vmDiskPath == "" {
+			deleteVM(ctx, dummyVM)
+			return "", fmt.Errorf("Unable to find the path for the disk on VM: %q with err: %+v", dummyVMName, err)
+		}
 
-	// Create virtual disk
-	task, err := virtualDiskManager.CreateVirtualDisk(ctx, vmDiskPath, dc, vmDiskSpec)
-	if err != nil {
-		return "", err
-	}
-	err = task.Wait(ctx)
-	if err != nil {
-		return "", err
-	}
+		dummyVMNodeName := vmNameToNodeName(dummyVMName)
+		// 2. Detach the disk from the dummy VM.
+		err = vs.DetachDisk(vmDiskPath, dummyVMNodeName)
+		if err != nil {
+			deleteVM(ctx, dummyVM)
+			glog.Errorf("Failed to detach the disk: %q from VM: %q with err: %+v", vmDiskPath, dummyVMName, err)
+			return "", fmt.Errorf("Failed to create the volume: %q with err: %+v", volumeOptions.Name, err)
+		}
+
+		// 3. Delete the dummy VM
+		err = deleteVM(ctx, dummyVM)
+		if err != nil {
+			glog.Errorf("Failed to delete the VM: %q with err: %+v", dummyVMName, err)
+			return "", fmt.Errorf("Failed to create the volume: %q with err: %+v", volumeOptions.Name, err)
+		}
+
+		kubeVolsPath := filepath.Clean(ds.Path(VolDir)) + "/"
+		// Create a kubevols directory in the datastore if one doesn't exist.
+		err = makeDirectoryInDatastore(vs.client, dc, kubeVolsPath, false)
+		if err != nil && err != ErrFileAlreadyExist {
+			glog.Errorf("Cannot create dir %#v. err %s", kubeVolsPath, err)
+			return "", err
+		}
+		glog.V(4).Infof("Created dir with path as %+q", kubeVolsPath)
+		destVolPath = kubeVolsPath + volumeOptions.Name + ".vmdk"
+
+		// 4. Move the Virtual Disk to Kubevols directory.
+		err = moveVirtualDisk(ctx, vs.client, dc, vmDiskPath, destVolPath)
+		if err != nil {
+			glog.Errorf("Failed to move the virtual disk from source: %+q to destination: %+q with err: %+v", vmDiskPath, destVolPath, err)
+			return "", fmt.Errorf("Failed to create the volume: %q with err: %+v", volumeOptions.Name, err)
```

Comment:
can we keep fmt.Errorf("Failed to create the volume: %q with err: %+v", volumeOptions.Name, err) in some error object and return on any failure. message is repeated several times.